lib/internal/cluster/round_robin_handle.js

```diff
@@ -10,115 +10,100 @@ const net = require('net');
-  assert(this.all.has(worker.id) === false);
-  this.all.set(worker.id, worker);
+  attachListener(fd, port, address, flags) {
+    if (fd >= 0)
+      this.server.listen({ fd });
+    else if (port >= 0) {
+      this.server.listen({
+        port,
+        host: address,
+        // Currently, net module only supports `ipv6Only` option in `flags`.
+        ipv6Only: Boolean(flags & constants.UV_TCP_IPV6ONLY),
+      });
+    } else
+      this.server.listen(address);
+  }
+  add(worker, send) {
+    assert(!this.all.has(worker.id));
+    this.all.set(worker.id, worker);
+    const done = () => {
+      if (this.handle.getsockname) {
+        const out = {};
+        this.handle.getsockname(out);
+        // TODO(bnoordhuis) Check err.
+        send(null, { sockname: out }, null);
+      } else {
+        send(null, null, null); // UNIX socket.
+      }
+      this.handoff(worker); // In case there are connections pending.
+    };
+    if (this.server === null)
+      return done();
+    // Still busy binding.
+    this.server.once('listening', done);
+    this.server.once('error', (err) => {
+      send(err.errno, null);
+    });
+  }
 
-  const done = () => {
-    if (this.handle.getsockname) {
-      const out = {};
-      this.handle.getsockname(out);
-      // TODO(bnoordhuis) Check err.
-      send(null, { sockname: out }, null);
-    } else {
-      send(null, null, null);  // UNIX socket.
+  remove(worker) {
+    const existed = this.all.delete(worker.id);
+    if (!existed)
+      return false;
+    this.free.delete(worker.id);
+    if (this.all.size !== 0)
+      return false;
+    for (const handle of this.handles) {
+      handle.close();
     }
-
-    this.handoff(worker);  // In case there are connections pending.
-  };
-
-  if (this.server === null)
-    return done();
-
-  // Still busy binding.
-  this.server.once('listening', done);
-  this.server.once('error', (err) => {
-    send(err.errno, null);
-  });
-};
-
-RoundRobinHandle.prototype.remove = function(worker) {
-  const existed = this.all.delete(worker.id);
-
-  if (!existed)
-    return false;
-
-  const index = this.free.indexOf(worker);
-
-  if (index !== -1)
-    this.free.splice(index, 1);
-
-  if (this.all.size !== 0)
-    return false;
-
-  for (const handle of this.handles) {
-    handle.close();
+    this.handles = [];
+    this.handle.close();
+    this.handle = null;
+    return true;
   }
-  this.handles = [];
-
-  this.handle.close();
-  this.handle = null;
-  return true;
-};
 
-RoundRobinHandle.prototype.distribute = function(err, handle) {
-  this.handles.push(handle);
-  const worker = this.free.shift();
-
-  if (worker)
-    this.handoff(worker);
-};
-
-RoundRobinHandle.prototype.handoff = function(worker) {
-  if (this.all.has(worker.id) === false) {
-    return;  // Worker is closing (or has closed) the server.
+  distribute(err, handle) {
+    this.handles.push(handle);
+    const worker = this.free.values().next().value;
```

Comment:
The previous implementation actually removed the worker from `.free()`, this will keep the worker in place.

Reply:
Can get the key and remove it from the map instead.